proposals/0000-namespace-specified-imports.rst

```diff
@@ -0,0 +1,656 @@
+
+* ``import M (data D(..))`` is syntactically valid, but not useful, as it is not
+  currently possible for identifiers in the data namespace to have children.
+  (We might imagine changing this e.g. for pattern synonym record fields, but
+  doing so is outside the scope of this proposal.)
+
+
+Deprecate use of ``pattern`` in import/export lists
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Since the ``data`` specifier introduced above subsumes uses of the ``pattern``
+keyword in import/export lists that are permitted under ``PatternSynonyms``, we
+propose a new warning ``-Wpattern-namespace-specifier`` that warns when the
+``pattern`` namespace specifier is used.
+
+Initially this warning will be added to ``-Wcompat``.  Three releases after this
+proposal is implemented, the warning will be added to ``-Wall``.
+
+We do not currently propose increasing the severity of the warning beyond
+``-Wall``, or removing support for ``pattern`` in import/export lists entirely,
+because the simplification to the compiler does not seem worth the backwards
+compatibility cost.
+
+
+Namespace-specified imports
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When the ``ExplicitNamespaces`` extension is enabled, the syntax of import
+declarations is extended to include a namespace specifier immediately after the
+module identifier.
+
+More concretely, in the grammar accepted by GHC, ::
+
+      importdecl -> 'import' [src] ['safe'] ['qualified'] [package] modid ['qualified'] ['as' modid] [impspec]
+
+is changed to ::
+
+      importdecl -> 'import' [src] ['safe'] ['qualified'] [package] modid [namespace] ['qualified'] ['as' modid] [impspec]
+
+      namespace  -> 'data'
+                 |  'type'
+
+With a namespace specified in the import, only identifiers belonging to the
+corresponding namespace will be brought into the scope, as if an explicit import
+list was given mentioning only those identifiers (with the namespace specifier
+on each item).
+
+If an import declaration uses both a namespace specifier and an explicit import
+list, the explicit import list may not mention a different namespace specifier,
+nor an identifier that is not available in the given namespace, otherwise a name
+resolution error will be reported.  It is allowed to redundantly specify the
+same namespace specifier on the import declaration and on an individual item.
+
+If an import declaration uses a namespace specifier but no explicit import list,
+it is not an error for the declaration to bring no names into scope,
+e.g. because the ``data`` specifier was used on a module that exports only type
+names. (GHC may of course warn that such an import is redundant.)
+
+Where an import has a namespace specifier, any occurrence of an ellipsis
+``(..)`` will be taken to refer only to identifiers in that namespace.  For
+example, ``import M type (T(..))`` will import only type-level names (so the
+ellipsis will refer to nothing if ``T`` is a normal algebraic datatype).
```

Comment:
The ellipsis will refer to nothing, but the `T` type will still be imported, right? (Maybe say: "so for a normal algebraic data type `T` only the type `T` is imported and ellipsis will refer to nothing")